internal/gcsx/client_readers/gcs_reader_test.go

```diff
@@ -269,3 +307,184 @@ func (t *gcsReaderTest) Test_ReadAt_ValidateReadType() {
 		})
 	}
 }
+
+func (t *gcsReaderTest) Test_ReadAt_PropagatesCancellation() {
+	// Set up a blocking reader
+	finishRead := make(chan struct{})
+	blocking := &blockingReader{c: finishRead}
+	rc := io.NopCloser(blocking)
+	// Assign it to the rangeReader
+	t.gcsReader.rangeReader.reader = &fake.FakeReader{ReadCloser: rc}
+	t.gcsReader.rangeReader.start = 0
+	t.gcsReader.rangeReader.limit = 2
+	// Track cancel invocation
+	cancelCalled := make(chan struct{})
+	t.gcsReader.rangeReader.cancel = func() { close(cancelCalled) }
+	// Controlled context
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	// Channel to track read completion
+	readReturned := make(chan struct{})
+	var err error
+
+	go func() {
+		_, err = t.gcsReader.ReadAt(ctx, make([]byte, 2), 0)
+
+		assert.Error(t.T(), err)
+
+		close(readReturned)
+	}()
+
+	// Wait a bit to ensure ReadAt is blocking
+	select {
+	case <-readReturned:
+		t.T().Fatal("Read returned early — cancellation did not propagate properly.")
+	case <-time.After(10 * time.Millisecond):
+		// OK: Still blocked
+	}
+	// Cancel the context to trigger cancellation
+	cancel()
+	// Expect rr.cancel to be called
+	select {
+	case <-cancelCalled:
+		// Pass
+	case <-time.After(100 * time.Millisecond):
+		t.T().Fatal("Expected rr.cancel to be called on ctx cancellation.")
+	}
+	// Unblock the reader so the read can complete
+	close(finishRead)
+	// Ensure read completes
+	select {
+	case <-readReturned:
+		// Pass
+	case <-time.After(100 * time.Millisecond):
+		t.T().Fatal("Expected read to return after unblocking.")
+	}
+}
+
+func (t *gcsReaderTest) Test_ReadInfo() {
+	t.object.Size = 10 * MiB
+	testCases := []struct {
+		name  string
+		start int64
+		size  int64
+	}{
+		{
+			name:  "startLessThanZero",
+			start: -1,
+			size:  10,
+		},
+		{
+			name:  "sizeLessThanZero",
+			start: 0,
+			size:  -1,
+		},
+		{
+			name:  "startGreaterThanObjectSize",
+			start: int64(t.object.Size + 1),
+			size:  int64(t.object.Size),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func() {
+			_, err := t.gcsReader.getReadInfo(tc.start, tc.size)
+
+			assert.Error(t.T(), err)
+		})
+	}
+}
+
+func (t *gcsReaderTest) Test_ReadInfo_Sequential() {
+	testCases := []struct {
+		name        string
+		start       int64
+		objectSize  uint64
+		expectedEnd int64
+	}{
+		{
+			name:        "ExactSizeRead", // start 0, object = 10MB
+			start:       0,
+			objectSize:  10 * MiB,
+			expectedEnd: 10 * MiB,
+		},
+		{
+			name:        "ReadSizeGreaterThanObjectSize", // start near end, should clamp to objectSize
+			start:       int64(10*MiB - 1),
+			objectSize:  10 * MiB,
+			expectedEnd: 10 * MiB,
+		},
+		{
+			name:        "ObjectSizeGreaterThanReadSize", // default read size applies
+			start:       0,
+			objectSize:  50 * MiB,
+			expectedEnd: int64(sequentialReadSizeInBytes),
```

Comment:
nit: Can we write the exact number (22 * MB)? Add a inline comment sequentialReadSizeInBytes